### ISSUE TITLE
Cache extent supports caching by on number of items

### ISSUE THREAD

**Issue by chunhtai** [flutter/flutter#128601]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Use case

Right now, you can set cache extent style by pixel or viewport. There should be a way to cache by number of items.

### Proposal

Find a way to support this use case. One idea is to add a new enum  to CacheExtentStyle.

**Comment by yjbanov** [comment]
Internal bug b/286441683

**Comment by Piinks** [comment]
I really like this idea. More context from the bug, this affects a11y when the content and the cache extent are such that you cannot traverse to the next item. Being able to specify the minimum number of items rendered outside of the viewport sounds like a much better was to ensure content can be traversable.

**Comment by Piinks** [comment]
Actually, I wonder if we should just do this by default. Here's my thinking:

This only applies to the individual slivers of the viewport that have multiple children. All slivers always lay out at least one child. So what we are really looking at are slivers like SliverList and SliverGrid, etc. that have this corner case where the extent of their laid out children fills the cacheExtent just right that we can't traverse to a next child outside of the viewport.

The viewport itself does not know about the individual children of the slivers, it just passes that cacheExtent info to the actual slivers when they are laid out.
Since this case does not apply to all slivers, then maybe we should just update the slivers with multiple children to always lay out one last child that is 100% not visible by default. (My understanding is the most common corner case of this issue is a large child is partially visible and consumes the full cacheExtent, so no next child is waiting outside of the viewport to change the focus to)

I could also see this issue not just making the content inaccessible, but also creating odd side effects - such as if the SliverList that does not have its next child ready in the extent has yet _another_ sliver following it, the focus traversal would break by skipping the rest of the leading SliverList's content. 

This feels like a really unfortunate gotcha for developers that we can probably just help them avoid all together. I don't know how many developers would readily understand what is happening in this scenario and easily discover a cache extent API to configure it just so. It feels pretty low level to me from a developer perspective, but I could be wrong.

@chunhtai what do you think of this?

**Comment by chunhtai** [comment]
I agree by default we should cache by number of item instead of some fixed pixel.

 > Since this case does not apply to all slivers, then maybe we should just update the slivers with multiple children to always lay out one last child that is 100% not visible by default.

from my experience, just one child may not be enough. I have seen several time that iOS struggle to keep up with newly added child after the scroll. If you keep swipe right fast, sometimes you would hear a end of the list sound. We should pick a number `n` for a smoother user experience.

Also we should be aware not every child has a semantics node, I think the easiest way will be expose the api to control the `n`, or we  can be smarter to keep building until we found one that have semantics node. I think the latter is possible, but will probably need to change the rendering and semantics logic quite a bit.
